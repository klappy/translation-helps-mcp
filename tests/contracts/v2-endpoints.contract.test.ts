/**
 * Contract Tests for v2 Endpoints
 *
 * These tests capture the exact behavior of our v2 endpoints.
 * They serve as a contract that must not be broken when we
 * connect real data sources.
 */

import { describe, expect, it } from "vitest";

const BASE_URL = "http://localhost:8174/api/v2";

// Helper to make requests
async function fetchEndpoint(
  path: string,
  params: Record<string, string> = {},
) {
  const url = new URL(`${BASE_URL}${path}`);
  Object.entries(params).forEach(([key, value]) => {
    url.searchParams.append(key, value);
  });

  const response = await fetch(url.toString());
  return {
    status: response.status,
    data: await response.json(),
    headers: Object.fromEntries(response.headers.entries()),
  };
}

describe("Scripture Endpoints Contract", () => {
  describe("fetch-scripture", () => {
    it("returns multiple translations for a reference", async () => {
      const result = await fetchEndpoint("/fetch-scripture", {
        reference: "John 3:16",
      });

      expect(result.status).toBe(200);
      expect(result.data).toMatchSnapshot("fetch-scripture-john-3-16");

      // Verify structure
      expect(result.data).toHaveProperty("scripture");
      expect(result.data).toHaveProperty("metadata");
      expect(result.data.scripture).toBeInstanceOf(Array);
      expect(result.data.scripture.length).toBeGreaterThan(0);

      // Verify each scripture item
      result.data.scripture.forEach((item: any) => {
        expect(item).toHaveProperty("text");
        expect(item).toHaveProperty("reference");
        expect(item).toHaveProperty("resource");
        expect(item).toHaveProperty("citation");
      });
    });

    it("filters by specific resource", async () => {
      const result = await fetchEndpoint("/fetch-scripture", {
        reference: "Genesis 1:1",
        resource: "ult",
      });

      expect(result.status).toBe(200);
      expect(result.data.scripture).toHaveLength(1);
      expect(result.data.scripture[0].resource).toBe("ult");
    });

    it("returns 404 for unknown reference", async () => {
      const result = await fetchEndpoint("/fetch-scripture", {
        reference: "Unknown 99:99",
      });

      expect(result.status).toBe(404);
      expect(result.data).toHaveProperty("error");
    });
  });

  describe("fetch-ult-scripture", () => {
    it("returns only ULT translation", async () => {
      const result = await fetchEndpoint("/fetch-ult-scripture", {
        reference: "Psalm 23:1",
      });

      expect(result.status).toBe(200);
      expect(result.data).toMatchSnapshot("fetch-ult-psalm-23-1");
      expect(result.data.scripture).toHaveLength(1);
      expect(result.data.scripture[0].resource).toBe("ult");
      expect(result.data.scripture[0].version).toBe("ULT");
    });
  });

  describe("fetch-ust-scripture", () => {
    it("returns only UST translation", async () => {
      const result = await fetchEndpoint("/fetch-ust-scripture", {
        reference: "Romans 8:28",
      });

      expect(result.status).toBe(200);
      expect(result.data).toMatchSnapshot("fetch-ust-romans-8-28");
      expect(result.data.scripture).toHaveLength(1);
      expect(result.data.scripture[0].resource).toBe("ust");
      expect(result.data.scripture[0].version).toBe("UST");
    });
  });
});

describe("Translation Helps Endpoints Contract", () => {
  describe("translation-notes", () => {
    it("returns notes for a reference", async () => {
      const result = await fetchEndpoint("/translation-notes", {
        reference: "John 3:16",
        language: "en",
        organization: "unfoldingWord",
      });

      expect(result.status).toBe(200);
      expect(result.data).toMatchSnapshot("translation-notes-john-3-16");
      expect(result.data).toHaveProperty("items");
      expect(result.data.items).toBeInstanceOf(Array);
    });
  });

  describe("translation-words", () => {
    it("returns word definitions for a reference", async () => {
      const result = await fetchEndpoint("/fetch-translation-words", {
        reference: "Genesis 1:1",
        language: "en",
      });

      expect(result.status).toBe(200);
      expect(result.data).toMatchSnapshot("translation-words-genesis-1-1");
      expect(result.data).toHaveProperty("items");
      expect(result.data.items).toBeInstanceOf(Array);
    });
  });
});

describe("Discovery Endpoints Contract", () => {
  describe("simple-languages", () => {
    it("returns list of available languages", async () => {
      const result = await fetchEndpoint("/simple-languages");

      expect(result.status).toBe(200);
      expect(result.data).toMatchSnapshot("languages-list");
      expect(result.data).toHaveProperty("items");
      expect(result.data.items).toBeInstanceOf(Array);
      expect(result.data.items.length).toBeGreaterThan(0);
    });
  });

  describe("get-available-books", () => {
    it("returns list of Bible books", async () => {
      const result = await fetchEndpoint("/get-available-books", {
        language: "en",
        organization: "unfoldingWord",
      });

      expect(result.status).toBe(200);
      expect(result.data).toMatchSnapshot("available-books-en");
      expect(result.data).toHaveProperty("items");
      expect(result.data.items).toBeInstanceOf(Array);
    });
  });
});

describe("Complex Endpoints Contract", () => {
  describe("get-context", () => {
    it("aggregates all resources for a reference", async () => {
      const result = await fetchEndpoint("/get-context", {
        reference: "John 3:16",
      });

      expect(result.status).toBe(200);
      expect(result.data).toMatchSnapshot("get-context-john-3-16");

      // Should have all resource types
      expect(result.data).toHaveProperty("scripture");
      expect(result.data).toHaveProperty("translationNotes");
      expect(result.data).toHaveProperty("translationWords");
      expect(result.data).toHaveProperty("translationQuestions");
      expect(result.data).toHaveProperty("translationAcademy");
      expect(result.data).toHaveProperty("crossReferences");
    });
  });

  describe("resource-recommendations", () => {
    it("provides intelligent recommendations", async () => {
      const result = await fetchEndpoint("/resource-recommendations", {
        reference: "John 3:16",
        userRole: "translator",
      });

      expect(result.status).toBe(200);
      expect(result.data).toMatchSnapshot(
        "recommendations-john-3-16-translator",
      );
      expect(result.data).toHaveProperty("items");
      expect(result.data.items).toBeInstanceOf(Array);

      // Each recommendation should have required fields
      result.data.items.forEach((item: any) => {
        expect(item).toHaveProperty("resourceType");
        expect(item).toHaveProperty("title");
        expect(item).toHaveProperty("reason");
        expect(item).toHaveProperty("confidence");
        expect(item).toHaveProperty("rank");
      });
    });
  });
});

describe("Error Handling Contract", () => {
  it("returns consistent error format for missing parameters", async () => {
    const result = await fetchEndpoint("/fetch-scripture", {});

    expect(result.status).toBe(400);
    expect(result.data).toHaveProperty("error");
    expect(result.data).toMatchSnapshot("error-missing-reference");
  });

  it("returns consistent error format for invalid parameters", async () => {
    const result = await fetchEndpoint("/fetch-scripture", {
      reference: "Not a valid reference!",
    });

    expect(result.status).toBe(400);
    expect(result.data).toHaveProperty("error");
    expect(result.data).toMatchSnapshot("error-invalid-reference");
  });
});

describe("Response Consistency Contract", () => {
  it("all endpoints return consistent metadata structure", async () => {
    const endpoints = [
      { path: "/fetch-scripture", params: { reference: "John 3:16" } },
      { path: "/translation-notes", params: { reference: "John 3:16" } },
      { path: "/simple-languages", params: {} },
      { path: "/get-available-books", params: { language: "en" } },
    ];

    for (const endpoint of endpoints) {
      const result = await fetchEndpoint(endpoint.path, endpoint.params);

      expect(result.status).toBe(200);
      expect(result.data).toHaveProperty("metadata");

      const metadata = result.data.metadata;
      expect(metadata).toHaveProperty("totalCount");

      // No diagnostic data in response body
      expect(metadata).not.toHaveProperty("xrayTrace");
      expect(metadata).not.toHaveProperty("_diagnostics");
      expect(metadata).not.toHaveProperty("_internal");
    }
  });
});
