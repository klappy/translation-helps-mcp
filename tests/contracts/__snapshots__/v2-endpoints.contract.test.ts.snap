// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Complex Endpoints Contract > get-context > aggregates all resources for a reference > get-context-john-3-16 1`] = `
{
  "crossReferences": [
    "Romans 5:8",
    "1 John 4:9-10",
    "John 10:28",
    "Romans 6:23",
  ],
  "language": "en",
  "metadata": {
    "aggregationTime": 111,
    "coverage": {
      "hasAcademy": true,
      "hasCrossReferences": true,
      "hasNotes": true,
      "hasQuestions": true,
      "hasScripture": true,
      "hasWords": true,
    },
    "language": "en",
    "organization": "unfoldingWord",
    "reference": "John 3:16",
    "resourcesFound": {
      "academy": 2,
      "crossReferences": 4,
      "notes": 2,
      "questions": 2,
      "scripture": true,
      "words": 3,
    },
    "source": "aggregated-context",
    "timestamp": "2025-08-12T23:03:24.398Z",
    "totalResources": 14,
  },
  "organization": "unfoldingWord",
  "reference": "John 3:16",
  "scripture": {
    "reference": "John 3:16",
    "text": "For God so loved the world, that he gave his only Son, that whoever believes in him should not perish but have eternal life.",
    "version": "ULT",
  },
  "translationAcademy": [
    {
      "id": "ta_metaphor",
      "relevance": "Understanding "perish" as spiritual death",
      "title": "Metaphor",
    },
    {
      "id": "ta_abstract_nouns",
      "relevance": "Translating concepts like "love" and "life"",
      "title": "Abstract Nouns",
    },
  ],
  "translationNotes": [
    {
      "content": "This is one of the most famous verses in the Bible, summarizing the gospel message.",
      "id": "tn_john_3_16_001",
      "noteType": "general",
      "quote": "For God so loved",
    },
    {
      "content": "The Greek word for "world" here refers to all humanity, not the physical earth.",
      "id": "tn_john_3_16_002",
      "noteType": "key-term",
      "quote": "the world",
    },
  ],
  "translationQuestions": [
    {
      "answer": "God gave his only Son.",
      "id": "tq_john_3_16_001",
      "question": "What did God give because of his love for the world?",
    },
    {
      "answer": "They will not perish but have eternal life.",
      "id": "tq_john_3_16_002",
      "question": "What happens to those who believe in God's Son?",
    },
  ],
  "translationWords": [
    {
      "definition": "To have a strong affection for someone; to care deeply.",
      "occurrence": "loved",
      "strongs": [
        "G25",
      ],
      "term": "love",
    },
    {
      "definition": "Life that never ends; spiritual life that continues after physical death.",
      "occurrence": "eternal life",
      "strongs": [
        "G166",
        "G2222",
      ],
      "term": "eternal-life",
    },
    {
      "definition": "To trust; to have faith in someone or something.",
      "occurrence": "believes",
      "strongs": [
        "G4100",
      ],
      "term": "believe",
    },
  ],
}
`;

exports[`Complex Endpoints Contract > resource-recommendations > provides intelligent recommendations > recommendations-john-3-16-translator 1`] = `
{
  "items": [
    {
      "confidence": 100,
      "priority": 0,
      "rank": 1,
      "reason": "This verse contains multiple theological key terms",
      "resourceType": "tw",
      "title": "Key Terms: Love, World, Eternal Life",
    },
    {
      "confidence": 90,
      "priority": 1,
      "rank": 2,
      "reason": "Critical for understanding cultural and historical context in Gospel narratives",
      "resourceType": "tn",
      "title": "Translation Notes",
    },
    {
      "confidence": 80,
      "priority": 2,
      "rank": 3,
      "reason": "Key terms like "Son of God", "Kingdom of Heaven" need careful translation",
      "resourceType": "tw",
      "title": "Translation Words",
    },
    {
      "confidence": 70,
      "priority": 3,
      "rank": 4,
      "reason": "Helps ensure key Gospel themes are clearly communicated",
      "resourceType": "tq",
      "title": "Translation Questions",
    },
    {
      "confidence": 60,
      "priority": 4,
      "rank": 5,
      "reason": "Articles on parables, metaphors, and narrative techniques",
      "resourceType": "ta",
      "title": "Translation Academy",
    },
  ],
  "metadata": {
    "categories": {
      "advanced": 0,
      "essential": 3,
      "helpful": 2,
      "specific": 1,
    },
    "confidenceRange": {
      "average": 80,
      "max": 100,
      "min": 60,
    },
    "genre": "gospel",
    "language": "en",
    "organization": "unfoldingWord",
    "reference": "John 3:16",
    "source": "recommendation-engine",
    "totalCount": 5,
    "totalRecommendations": 5,
    "userRole": "translator",
  },
}
`;

exports[`Discovery Endpoints Contract > get-available-books > returns list of Bible books > available-books-en 1`] = `
{
  "items": [],
  "metadata": {
    "circuitBreakerState": "CLOSED",
    "language": "en",
    "organization": "unfoldingWord",
    "resource": "ult",
    "source": "DCS API",
    "totalCount": 0,
  },
}
`;

exports[`Discovery Endpoints Contract > simple-languages > returns list of available languages > languages-list 1`] = `
{
  "items": [
    {
      "code": "en",
      "direction": "ltr",
      "name": "English",
      "resources": {
        "ta": true,
        "tn": true,
        "tq": true,
        "tw": true,
        "ult": true,
        "ust": true,
      },
    },
    {
      "code": "es",
      "direction": "ltr",
      "name": "Español",
      "resources": {
        "ta": false,
        "tn": true,
        "tq": false,
        "tw": true,
        "ult": true,
        "ust": true,
      },
    },
    {
      "code": "fr",
      "direction": "ltr",
      "name": "Français",
      "resources": {
        "ta": false,
        "tn": false,
        "tq": false,
        "tw": false,
        "ult": true,
        "ust": true,
      },
    },
    {
      "code": "ar",
      "direction": "rtl",
      "name": "العربية",
      "resources": {
        "ta": false,
        "tn": false,
        "tq": false,
        "tw": false,
        "ult": true,
        "ust": false,
      },
    },
  ],
  "metadata": {
    "hasMore": false,
    "totalCount": 4,
  },
}
`;

exports[`Error Handling Contract > returns consistent error format for invalid parameters > error-invalid-reference 1`] = `
{
  "details": [
    "Parameter reference is invalid",
  ],
  "error": "Invalid parameters",
  "status": 400,
}
`;

exports[`Error Handling Contract > returns consistent error format for missing parameters > error-missing-reference 1`] = `
{
  "details": [
    "Missing required parameter: reference",
  ],
  "error": "Invalid parameters",
  "status": 400,
}
`;

exports[`Scripture Endpoints Contract > fetch-scripture > returns multiple translations for a reference > fetch-scripture-john-3-16 1`] = `
{
  "citation": {
    "format": "json",
    "foundResources": [
      "ult",
      "ust",
      "t4t",
      "ueb",
    ],
    "language": "en",
    "organization": "unfoldingWord",
    "reference": "John 3:16",
    "requestedResources": [
      "ult",
      "ust",
      "t4t",
      "ueb",
    ],
  },
  "language": "en",
  "metadata": {
    "format": "json",
    "foundResources": [
      "ult",
      "ust",
      "t4t",
      "ueb",
    ],
    "language": "en",
    "organization": "unfoldingWord",
    "reference": "John 3:16",
    "requestedResources": [
      "ult",
      "ust",
      "t4t",
      "ueb",
    ],
    "resources": [
      "ult",
      "ust",
      "t4t",
      "ueb",
    ],
    "totalCount": 4,
  },
  "organization": "unfoldingWord",
  "scripture": [
    {
      "citation": "John 3:16 (ULT)",
      "copyright": "Public Domain",
      "direction": "ltr",
      "language": "en",
      "organization": "unfoldingWord",
      "reference": "John 3:16",
      "resource": "ult",
      "resourceInfo": {
        "description": "A form-centric translation that closely follows the Hebrew and Greek texts",
        "language": "en",
        "name": "unfoldingWord Literal Text",
        "textDirection": "ltr",
      },
      "text": "For God loved the world in this way: he gave his one and only Son, so that everyone who believes in him will not perish but will have eternal life.",
      "version": "ULT",
    },
    {
      "citation": "John 3:16 (UST)",
      "copyright": "Public Domain",
      "direction": "ltr",
      "language": "en",
      "organization": "unfoldingWord",
      "reference": "John 3:16",
      "resource": "ust",
      "resourceInfo": {
        "description": "A meaning-centric translation that expresses the meanings of the Hebrew and Greek texts",
        "language": "en",
        "name": "unfoldingWord Simplified Text",
        "textDirection": "ltr",
      },
      "text": "God loved all the people in the world in this way: He gave his only Son to die for them. The result is that everyone who trusts in him will not perish. Instead, they will live forever.",
      "version": "UST",
    },
    {
      "citation": "John 3:16 (T4T)",
      "copyright": "Public Domain",
      "direction": "ltr",
      "language": "en",
      "organization": "unfoldingWord",
      "reference": "John 3:16",
      "resource": "t4t",
      "resourceInfo": {
        "description": "A meaning-based translation designed to help translators understand the text",
        "language": "en",
        "name": "Translation for Translators",
        "textDirection": "ltr",
      },
      "text": "God loved us people in the world so much that he gave his only Son as a sacrifice for us, in order that everyone who believes in him would not be separated from God forever. Instead, they would have eternal life.",
      "version": "T4T",
    },
    {
      "citation": "John 3:16 (UEB)",
      "copyright": "Public Domain",
      "direction": "ltr",
      "language": "en",
      "organization": "unfoldingWord",
      "reference": "John 3:16",
      "resource": "ueb",
      "resourceInfo": {
        "description": "An open-licensed English translation designed for clarity",
        "language": "en",
        "name": "Unlocked English Bible",
        "textDirection": "ltr",
      },
      "text": "For God so loved the world, that he gave his only Son, that whoever believes in him should not perish but have eternal life.",
      "version": "UEB",
    },
  ],
}
`;

exports[`Scripture Endpoints Contract > fetch-ult-scripture > returns only ULT translation > fetch-ult-psalm-23-1 1`] = `
{
  "citation": {
    "checkingLevel": "3",
    "language": "en",
    "organization": "unfoldingWord",
    "reference": "Psalm 23:1",
    "resource": "ult",
    "translator": "unfoldingWord",
  },
  "language": "en",
  "metadata": {
    "checkingLevel": "3",
    "language": "en",
    "organization": "unfoldingWord",
    "reference": "Psalm 23:1",
    "resource": "ult",
    "resources": [
      "ult",
    ],
    "totalCount": 1,
    "translator": "unfoldingWord",
  },
  "organization": "unfoldingWord",
  "scripture": [
    {
      "citation": "Psalm 23:1 (ULT)",
      "copyright": "Public Domain",
      "direction": "ltr",
      "language": "en",
      "organization": "unfoldingWord",
      "reference": "Psalm 23:1",
      "resource": "ult",
      "resourceInfo": {
        "abbreviation": "ULT",
        "checking_level": "3",
        "copyright": "Public Domain",
        "description": "A form-centric translation that closely follows the Hebrew and Greek texts",
        "language": "en",
        "licenseUrl": "https://creativecommons.org/publicdomain/mark/1.0/",
        "name": "unfoldingWord Literal Text",
        "publisher": "unfoldingWord",
        "textDirection": "ltr",
        "version": "37",
      },
      "text": "A psalm of David. Yahweh is my shepherd; I will lack nothing.",
      "version": "ULT",
    },
  ],
}
`;

exports[`Scripture Endpoints Contract > fetch-ust-scripture > returns only UST translation > fetch-ust-romans-8-28 1`] = `
{
  "citation": {
    "checkingLevel": "3",
    "language": "en",
    "organization": "unfoldingWord",
    "readingLevel": "7th grade",
    "reference": "Romans 8:28",
    "resource": "ust",
    "translator": "unfoldingWord",
  },
  "language": "en",
  "metadata": {
    "checkingLevel": "3",
    "language": "en",
    "organization": "unfoldingWord",
    "readingLevel": "7th grade",
    "reference": "Romans 8:28",
    "resource": "ust",
    "resources": [
      "ust",
    ],
    "totalCount": 1,
    "translator": "unfoldingWord",
  },
  "organization": "unfoldingWord",
  "scripture": [
    {
      "citation": "Romans 8:28 (UST)",
      "copyright": "Public Domain",
      "direction": "ltr",
      "language": "en",
      "organization": "unfoldingWord",
      "reference": "Romans 8:28",
      "resource": "ust",
      "resourceInfo": {
        "abbreviation": "UST",
        "checking_level": "3",
        "copyright": "Public Domain",
        "description": "A meaning-centric translation that expresses the meanings of the Hebrew and Greek texts",
        "language": "en",
        "licenseUrl": "https://creativecommons.org/publicdomain/mark/1.0/",
        "name": "unfoldingWord Simplified Text",
        "publisher": "unfoldingWord",
        "reading_level": "7th grade",
        "textDirection": "ltr",
        "version": "38",
      },
      "text": "And we know that for those who love God, he works to use all things for their good. This is true for those whom he has chosen to be his people.",
      "version": "UST",
    },
  ],
}
`;

exports[`Translation Helps Endpoints Contract > translation-notes > returns notes for a reference > translation-notes-john-3-16 1`] = `
{
  "items": [
    {
      "id": "tn003",
      "noteType": "general",
      "reference": "John 3:16",
      "supportReference": "rc://*/ta/man/translate/figs-explicit",
      "text": "God's love for the world is the motivation for his giving his Son.",
    },
    {
      "id": "tn004",
      "noteType": "key-term",
      "reference": "John 3:16",
      "term": "world",
      "text": "The word "world" here refers to all people.",
    },
  ],
  "language": "en",
  "metadata": {
    "language": "en",
    "organization": "unfoldingWord",
    "source": "TN",
    "totalCount": 2,
  },
  "organization": "unfoldingWord",
  "reference": "John 3:16",
}
`;

exports[`Translation Helps Endpoints Contract > translation-words > returns word definitions for a reference > translation-words-genesis-1-1 1`] = `
{
  "items": [],
  "language": "en",
  "metadata": {
    "language": "en",
    "organization": "unfoldingWord",
    "source": "TW",
    "totalCount": 0,
  },
  "organization": "unfoldingWord",
  "reference": "Genesis 1:1",
}
`;
